File: GGAndroid_2.0/GGCore/GGVMCore/src/main/java/io/ganguo/vmodel/adapter/ViewModelPagerAdapter.java
package io.ganguo.vmodel.adapter;

import android.support.annotation.NonNull;
import android.support.v4.view.PagerAdapter;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import io.ganguo.utils.util.Collections;
import io.ganguo.vmodel.BaseViewModel;
import io.ganguo.vmodel.ViewModelHelper;

/**
 * update by leo on 2018/07/07.
 * PagerAdapter - 支持DataBinding及ViewModel
 */
public class ViewModelPagerAdapter extends PagerAdapter {
    private List<BaseViewModel> pageViewModels = new ArrayList<>();
    private BaseViewModel viewModelParent;

    public ViewModelPagerAdapter(@NonNull BaseViewModel viewModelParent) {
        this.viewModelParent = viewModelParent;
    }

    public ViewModelPagerAdapter(@NonNull BaseViewModel parent, List<? extends BaseViewModel> items) {
        this.pageViewModels = (List<BaseViewModel>) items;
        this.viewModelParent = parent;
    }

    @Override
    public int getCount() {
        return Collections.isEmpty(pageViewModels) ? 0 : pageViewModels.size();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        BaseViewModel viewModel = getPageViewModels().get(position);
        ViewModelHelper.bind(container, viewModelParent, viewModel);
        return viewModel.getRootView();
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
        pageViewModels.get(position).onDestroy();
    }

    public List<BaseViewModel> getPageViewModels() {
        return pageViewModels;
    }

    public ViewModelPagerAdapter setPageViewModels(List<BaseViewModel> pageViewModels) {
        this.pageViewModels = pageViewModels;
        return this;
    }

    /**
     * function：add ViewModel
     *
     * @param pageVModel
     */
    public void addViewModel(BaseViewModel pageVModel) {
        pageViewModels.add(pageVModel);
    }

    /**
     * function：get ViewModel
     *
     * @param position
     * @return BaseViewModel
     */
    public BaseViewModel getPageViewModel(int position) {
        if (position >= getCount()) {
            return null;
        }
        return pageViewModels.get(position);
    }

}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/core/adapter/ViewModelPagerAdapter.kt
package io.ganguo.viewmodel.core.adapter

import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.ViewModelHelper
import io.ganguo.viewmodel.core.viewinterface.ViewInterface
import java.util.*


/**
 * <pre>
 *    @author : leo
 *    time   : 2020/05/23
 *    desc   : Adapter DiffUtil Comparator
 * </pre>
 * @property pageViewModels MutableList<BaseViewModel<*>>
 * @property viewModelParent BaseViewModel<*>
 */
class ViewModelPagerAdapter : PagerAdapter {
    var pageViewModels: MutableList<BaseViewModel<*>> = ArrayList()
    private var viewModelParent: BaseViewModel<*>

    constructor(viewModelParent: BaseViewModel<*>) {
        this.viewModelParent = viewModelParent
    }

    constructor(parent: BaseViewModel<*>, items: MutableList<BaseViewModel<*>>) {
        pageViewModels.addAll(items)
        viewModelParent = parent
    }

    /**
     * get page size
     * @return Int
     */
    override fun getCount(): Int {
        return pageViewModels.size
    }

    /**
     * equals item
     * @param view View
     * @param item Any
     * @return Boolean
     */
    override fun isViewFromObject(view: View, item: Any): Boolean {
        return view === item
    }

    /**
     * instantiate Item and binding data
     * @param container ViewGroup
     * @param position Int
     * @return Any
     */
    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val viewModel = pageViewModels[position]
        ViewModelHelper.bind(container, viewModelParent, viewModel)
        return viewModel.rootView
    }

    /**
     * Item destroy
     * @param container ViewGroup
     * @param position Int
     * @param item Any
     */
    override fun destroyItem(container: ViewGroup, position: Int, item: Any) {
        container.removeView(item as View)
        pageViewModels[position].onDestroy()
    }


    /**
     * Set up a new data source
     * @param pageViewModels List<BaseViewModel<*>>
     * @return ViewModelPagerAdapter
     */
    fun setPageViewModels(pageViewModels: List<BaseViewModel<*>>): ViewModelPagerAdapter {
        this.pageViewModels.clear()
        this.pageViewModels.addAll(pageViewModels)
        return this
    }

    /**
     * add ViewModel
     * @param pageVModel BaseViewModel<ViewInterface<*>>
     */
    fun addViewModel(pageVModel: BaseViewModel<ViewInterface<*>>) {
        pageViewModels.add(pageVModel)
    }

    /**
     * get Page ViewModel
     * @param position Int
     * @return BaseViewModel<*>
     */
    fun getPageViewModel(position: Int): BaseViewModel<*> {
        return pageViewModels[position]
    }
}